tftp-hpa-0.40/tftp/tftp.1.in
.\" -*- nroff -*- --------------------------------------------------------- *
.\" $Id: tftp.1.in,v 1.3 2004/01/08 20:47:00 hpa Exp $
.\"  
.\" Copyright (c) 1990, 1993, 1994
.\"     The Regents of the University of California.  All rights reserved.
.\"
.\" Copyright 2001 H. Peter Anvin - All Rights Reserved
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"----------------------------------------------------------------------- */
.TH TFTP 1 "2 February 2003" "tftp-hpa @@VERSION@@" "User's Manual"
.SH NAME
.B tftp
\- IPv4 Trivial File Transfer Protocol client
.SH SYNOPSIS
.B tftp
[ \fIoptions...\fP ]
[\fIhost\fP [\fIport\fP]]
[\fB\-c\fP \fIcommand\fP]
.br
.SH DESCRIPTION
.B tftp
is a client for the IPv4 Trivial file Transfer Protocol, which can be
used to transfer files to and from remote machines, including some
very minimalistic, usually embedded, systems.  The remote
.I host
may be specified on the command line, in which case
.B tftp
uses
.I host
as the default host for future transfers (see the
.B connect
command below.)
.SH OPTIONS
.TP
\fB\-c\fP \fIcommand\fP
Execute \fIcommand\fP as if it had been entered on the tftp prompt.
Must be specified last on the command line.
.TP
\fB\-m\fP \fImode\fP
Set the default transfer mode to \fImode\fP.  This is usually used with \-c.
.TP
.B \-v
Default to verbose mode.
.TP
.B \-V
Print the version number and configuration to standard output, then
exit gracefully.
.SH COMMANDS
Once
.B tftp
is running, it issues the prompt
\f(CWtftp>\fP
and recognizes the following commands:
.TP
\fB?\fP \fIcommand-name...\fP
.TP
\fBhelp\fP \fIcommand-name...\fP
Print help information
.TP
.B ascii
Shorthand for
.BR "mode ascii" .
.TP
.B binary
Shorthand for
.BR "mode binary" .
.TP
\fBconnect\fP \fIhost [port]\fP
Set the
.I host
(and optionally
.IR port )
for transfers.  Note that the TFTP protocol, unlike the FTP protocol,
does not maintain connections between transfers; thus, the
.B connect
command does not actually create a connection, but merely remembers
what host is to be used for transfers.  You do not have to use the
.B connect
command; the remote host can be specified as part of the
.B get
or
.B put
commands.
.TP
\fBget\fP \fIfile\fP
.sp -.6l
.TP
\fBget\fP \fIremotefile localfile\fP
.sp -.6l
.TP
\fBget\fP \fIfile1 file2 file3...\fP
Get a file or set of files from the specified sources.  A remote
filename can be in one of two forms: a plain filename on the remote
host, if the host has already been specified, or a string of the form
.I "host:filename"
to specify both a host and filename at the same time.  If the latter
form is used, the last hostname specified becomes the default for
future transfers.
.TP
\fBmode\fP \fItransfer-mode\fP
Specify the mode for transfers;
.I transfer-mode
may be one of
.B ascii
(or
.BR netascii )
or
.B binary
(or
.BR octet .)
The default is
.BR ascii .
.TP
\fBput\fP \fIfile\fP
.sp -.6l
.TP
\fBput\fP \fIlocalfile remotefile\fP
.sp -.6l
.TP
\fBput\fP \fIfile1 file2 file3... remote-directory\fP
Put a file or set of files to the specified remote file or directory.
The destination can be in one of two forms: a filename on the remote
host, if the host has already been specified, or a string of the form
.I "host:filename"
to specify both a host and filename at the same time.  If the latter
form is used, the hostname specified becomes the default for future
transfers.  If the remote-directory form is used, the remote host is
assumed to be a UNIX system or another system using
.B /
as directory separator.
.TP
.B quit
Exit
.BR tftp .
End-of-file will also exit.
.TP
\fBrexmt\fP \fIretransmission-timeout\fP
Set the per-packet retransmission timeout, in seconds.
.TP
.B status
Show current status.
.TP
\fBtimeout\fP \fItotal-transmission-timeout\fP
Set the total transmission timeout, in seconds.
.TP
.B trace
Toggle packet tracing (a debugging feature.)
.TP
.B verbose
Toggle verbose mode.
.SH "NOTES"
The TFTP protocol provides no provisions for authentication or
security.  Therefore, the remote server will probably implement some
kinds of access restriction or firewalling.  These access restrictions
are likely to be site- and server-specific.
.SH "AUTHOR"
This version of
.B tftp
is maintained by H. Peter Anvin <hpa@zytor.com>.  It was derived from,
but has substantially diverged from, an OpenBSD source base, with
added patches by Markus Gutschke and Gero Kulhman.
.SH "SEE ALSO"
.BR tftpd (8).


noflushd/NEWS
* noflushd 2.7.5 (09. May 2005)

  - Fix handling of raid and similar meta devices on 2.6 kernels.

* noflushd 2.7.4 (08. Aug 2004)

  - Fix lousy spindown times on 2.6 kernels.
  - Fix SCSI support on 2.6 kernels.
  - Fix pid detection in generic startup script.
  - Document quirks with hotplugged drives.

* noflushd 2.7.3 (09. Apr 2004)

  - Fix race condition that might leave kupdated/pdflushd disabled
    aftern noflushd has terminated.
  - Fix spindown when all disks are auto-detected.

* noflushd 2.7.2 (16. Mar 2004)

  - A change in 2.7.1 erroneously flushed the buffer cache before
    spindown, causing frequent spinups. This change has been reverted.
    
* noflushd 2.7.1 (07. Mar 2004)

  - Fix immediate spinup on 2.6 kernels.
  - Elevate SCSI support from compile-time to command line option.
  - Bug fixes in the disk statistics parser for 2.6 kernels.
  - Ship kernel patch to track down which processes cause spinups.

* noflushd 2.7 (20. Jan 2004)

  - Support Linux kernel 2.6.x.
  - Default to monitorring all disks unless specific disks are given
    on the command line. This change should yield a sane default
    behaviour for a majority of users without requiring explicit
    configuration.

* noflushd 2.6.3 (13. May 2002)

  - Bugfix for arm and other archs with weird /proc/partitions layout.

* noflushd 2.6.2 (20. Apr 2002)

  - Add IDE raid support.
  - Add timeout logging script.

* noflushd 2.6.1 (03. Feb. 2002)

  - Fix BUG when using noflushd with disks on second IDE channel on 2.4
    kernels.

* noflushd 2.6 (22. Jan 2002)

  - Document noflushd's suckage with any journaling fs, including ext3.
  - New feature prevents spindown if machine is in interactive use.
  - Add hack fix some '0 spindown time' uglies.

* noflushd 2.5 (06. Aug 2001)

  - Make sure hotplugged disks get synced properly.
  - Only sync present partitions. (Fixes high CPU usage bug on devfs
    systems.)
  - Generic startup skript works on latest 2.4 kernels.

* noflushd 2.4 (13. Mar 2001)

  - Timeout lists are back to a working state.
  - Support very latest 2.4 kernels.

* noflushd 2.3 (12. Jan 2001)

  - Support latest 2.4 kernels. (Leaves kernels 2.3.99pre8 to 2.4.0-test8
    unsupported.)
  - Fix multi-disk support with long bdflush timeouts.

* noflushd 2.2 (22. Dec 2000)

  - Fixed bug in spindown code. Now actually spins down. Duh!

* noflushd 2.1 (20. Dec 2000)

  - Disks get synced prior to spindown.
  - Improved devfs support.

* noflushd 2.0 (14. Nov 2000)

  - noflushd now comes with its own syncer that takes over the task from
    kupdate when the first read-write mounted disk is spun down. This
    enhances spindown times in a multidisk environment at the expense of
    less smooth and more bursty background flushing when the noflushd 
    syncer is in use.

  - Conflicting entries in 2.2-style /proc/stat (e.g. /dev/sda and /dev/hda)
    won't cause noflushd to give up anymore. Instead, noflushd will treat
    both disks pretty much like one, i.e. they're either both up or spun
    down, no matter if activity only goes to one of them.

  - Compilation environment switched to automake/autoconf.

  - When installing via 'make install', noflushd is no longer added to
    any runlevel. 

  - noflushd moved to SourceForge. The project homepage is located at
    http://noflushd.sourceforge.net, a mailing list is available via
    noflushd-devel@lists.sourceforge.net. Visit the homepage for 
    subscription information.



tftp-hpa-0.40/tftp/tftp.c
/* $Id: tftp.c,v 1.17 2004/01/08 20:47:00 hpa Exp $ */

/* $OpenBSD: tftp.c,v 1.4 1997/08/06 06:43:45 deraadt Exp $	*/
/* $NetBSD: tftp.c,v 1.5 1995/04/29 05:55:25 cgd Exp $	*/

/*
 * Copyright (c) 1983, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "tftpsubs.h"

#ifndef lint
/* static char sccsid[] = "@(#)tftp.c	8.1 (Berkeley) 6/6/93"; */
/* static char rcsid[] = "$OpenBSD: tftp.c,v 1.4 1997/08/06 06:43:45 deraadt Exp $"; */
static const char *rcsid UNUSED =
"tftp-hpa $Id: tftp.c,v 1.17 2004/01/08 20:47:00 hpa Exp $";
#endif /* not lint */

/* Many bug fixes are from Jim Guyton <guyton@rand-unix> */

/*
 * TFTP User Program -- Protocol Machines
 */
#include "extern.h"

extern  struct sockaddr_in peeraddr;	/* filled in by main */
extern  int     f;			/* the opened socket */
extern  int     trace;
extern  int     verbose;
extern  int     rexmtval;
extern  int     maxtimeout;

#define PKTSIZE    SEGSIZE+4
char    ackbuf[PKTSIZE];
int	timeout;
sigjmp_buf	toplevel;
sigjmp_buf	timeoutbuf;

static void nak(int, const char *);
static int makerequest(int, const char *, struct tftphdr *, const char *);
static void printstats(const char *, unsigned long);
static void startclock(void);
static void stopclock(void);
static void timer(int);
static void tpacket(const char *, struct tftphdr *, int);

/*
 * Send the requested file.
 */
void
tftp_sendfile(int fd, const char *name, const char *mode)
{
	struct tftphdr *ap;	   /* data and ack packets */
	struct tftphdr *dp;
	int n;
	volatile int is_request;
	volatile u_short block;
	volatile int size, convert;
	volatile off_t amount;
	struct sockaddr_in from;
	int fromlen;
	FILE *file;
	u_short ap_opcode, ap_block;

	startclock();		/* start stat's clock */
	dp = r_init();		/* reset fillbuf/read-ahead code */
	ap = (struct tftphdr *)ackbuf;
	convert = !strcmp(mode, "netascii");
	file = fdopen(fd, convert ? "rt" : "rb");
	block = 0;
	is_request = 1;		/* First packet is the actual WRQ */
	amount = 0;

	bsd_signal(SIGALRM, timer);
	do {
		if (is_request) {
			size = makerequest(WRQ, name, dp, mode) - 4;
		} else {
		/*	size = read(fd, dp->th_data, SEGSIZE);	 */
			size = readit(file, &dp, convert);
			if (size < 0) {
				nak(errno + 100, NULL);
				break;
			}
			dp->th_opcode = htons((u_short)DATA);
			dp->th_block = htons((u_short)block);
		}
		timeout = 0;
		(void) sigsetjmp(timeoutbuf,1);

		if (trace)
			tpacket("sent", dp, size + 4);
		n = sendto(f, dp, size + 4, 0,
		    (struct sockaddr *)&peeraddr, sizeof(peeraddr));
		if (n != size + 4) {
			perror("tftp: sendto");
			goto abort;
		}
		read_ahead(file, convert);
		for ( ; ; ) {
			alarm(rexmtval);
			do {
				fromlen = sizeof(from);
				n = recvfrom(f, ackbuf, sizeof(ackbuf), 0,
				    (struct sockaddr *)&from, &fromlen);
			} while (n <= 0);
			alarm(0);
			if (n < 0) {
				perror("tftp: recvfrom");
				goto abort;
			}
			peeraddr.sin_port = from.sin_port;	/* added */
			if (trace)
				tpacket("received", ap, n);
			/* should verify packet came from server */
			ap_opcode = ntohs((u_short)ap->th_opcode);
			ap_block = ntohs((u_short)ap->th_block);
			if (ap_opcode == ERROR) {
				printf("Error code %d: %s\n", ap_block,
					ap->th_msg);
				goto abort;
			}
			if (ap_opcode == ACK) {
				int j;

				if (ap_block == block) {
					break;
				}
				/* On an error, try to synchronize
				 * both sides.
				 */
				j = synchnet(f);
				if (j && trace) {
					printf("discarded %d packets\n",
							j);
				}
				/*
				 * RFC1129/RFC1350: We MUST NOT re-send the DATA
				 * packet in response to an invalid ACK.  Doing so
				 * would cause the Sorcerer's Apprentice bug.
				 */
			}
		}
		if ( !is_request )
			amount += size;
		is_request = 0;
		block++;
	} while (size == SEGSIZE || block == 1);
abort:
	fclose(file);
	stopclock();
	if (amount > 0)
		printstats("Sent", amount);
}

/*
 * Receive a file.
 */
void
tftp_recvfile(int fd, const char *name, const char *mode)
{
	struct tftphdr *ap;
	struct tftphdr *dp;
	int n;
	volatile u_short block;
	volatile int size, firsttrip;
	volatile unsigned long amount;
	struct sockaddr_in from;
	int fromlen;
	FILE *file;
	volatile int convert;		/* true if converting crlf -> lf */
	u_short dp_opcode, dp_block;

	startclock();
	dp = w_init();
	ap = (struct tftphdr *)ackbuf;
	convert = !strcmp(mode, "netascii");
	file = fdopen(fd, convert ?"wt":"wb");
	block = 1;
	firsttrip = 1;
	amount = 0;

	bsd_signal(SIGALRM, timer);
	do {
		if (firsttrip) {
			size = makerequest(RRQ, name, ap, mode);
			firsttrip = 0;
		} else {
			ap->th_opcode = htons((u_short)ACK);
			ap->th_block = htons((u_short)block);
			size = 4;
			block++;
		}
		timeout = 0;
		(void) sigsetjmp(timeoutbuf,1);
send_ack:
		if (trace)
			tpacket("sent", ap, size);
		if (sendto(f, ackbuf, size, 0, (struct sockaddr *)&peeraddr,
		    sizeof(peeraddr)) != size) {
			alarm(0);
			perror("tftp: sendto");
			goto abort;
		}
		write_behind(file, convert);
		for ( ; ; ) {
			alarm(rexmtval);
			do  {
				fromlen = sizeof(from);
				n = recvfrom(f, dp, PKTSIZE, 0,
				    (struct sockaddr *)&from, &fromlen);
			} while (n <= 0);
			alarm(0);
			if (n < 0) {
				perror("tftp: recvfrom");
				goto abort;
			}
			peeraddr.sin_port = from.sin_port;	/* added */
			if (trace)
				tpacket("received", dp, n);
			/* should verify client address */
			dp_opcode = ntohs((u_short)dp->th_opcode);
			dp_block = ntohs((u_short)dp->th_block);
			if (dp_opcode == ERROR) {
			  printf("Error code %d: %s\n", dp_block, dp->th_msg);
			  goto abort;
			}
			if (dp_opcode == DATA) {
				int j;

				if (dp_block == block) {
					break;		/* have next packet */
				}
				/* On an error, try to synchronize
				 * both sides.
				 */
				j = synchnet(f);
				if (j && trace) {
					printf("discarded %d packets\n", j);
				}
				if (dp_block == (block-1)) {
					goto send_ack;	/* resend ack */
				}
			}
		}
	/*	size = write(fd, dp->th_data, n - 4); */
		size = writeit(file, &dp, n - 4, convert);
		if (size < 0) {
			nak(errno + 100, NULL);
			break;
		}
		amount += size;
	} while (size == SEGSIZE);
abort:						/* ok to ack, since user */
	ap->th_opcode = htons((u_short)ACK);	/* has seen err msg */
	ap->th_block = htons((u_short)block);
	(void) sendto(f, ackbuf, 4, 0, (struct sockaddr *)&peeraddr,
	    sizeof(peeraddr));
	write_behind(file, convert);		/* flush last buffer */
	fclose(file);
	stopclock();
	if (amount > 0)
		printstats("Received", amount);
}

static int
makerequest(int request, const char *name,
	    struct tftphdr *tp, const char *mode)
{
	char *cp;

	tp->th_opcode = htons((u_short)request);
	cp = (char *) &(tp->th_stuff);
	strcpy(cp, name);
	cp += strlen(name);
	*cp++ = '\0';
	strcpy(cp, mode);
	cp += strlen(mode);
	*cp++ = '\0';
	return (cp - (char *)tp);
}

static const char * const errmsgs[] =
{
  "Undefined error code", 			/* 0 - EUNDEF */
  "File not found",				/* 1 - ENOTFOUND */
  "Access denied",				/* 2 - EACCESS */
  "Disk full or allocation exceeded", 		/* 3 - ENOSPACE */
  "Illegal TFTP operation",			/* 4 - EBADOP */
  "Unknown transfer ID",			/* 5 - EBADID */
  "File already exists",			/* 6 - EEXISTS */
  "No such user",				/* 7 - ENOUSER */
  "Failure to negotiate RFC2347 options" 	/* 8 - EOPTNEG */
};
#define ERR_CNT (sizeof(errmsgs)/sizeof(const char *))

/*
 * Send a nak packet (error message).
 * Error code passed in is one of the
 * standard TFTP codes, or a UNIX errno
 * offset by 100.
 */
static void
nak(int error, const char *msg)
{
  struct tftphdr *tp;
  int length;
  
  tp = (struct tftphdr *)ackbuf;
  tp->th_opcode = htons((u_short)ERROR);
  tp->th_code = htons((u_short)error);

  if ( error >= 100 ) {
    /* This is a Unix errno+100 */
    if ( !msg )
      msg = strerror(error - 100);
    error = EUNDEF;
  } else {
    if ( (unsigned)error >= ERR_CNT )
      error = EUNDEF;
    
    if ( !msg )
      msg = errmsgs[error];
  }

  tp->th_code = htons((u_short)error);

  length = strlen(msg)+1;
  memcpy(tp->th_msg, msg, length);
  length += 4;			/* Add space for header */

  if (trace)
    tpacket("sent", tp, length);
  if (sendto(f, ackbuf, length, 0, (struct sockaddr *)&peeraddr,
	     sizeof(peeraddr)) != length)
    perror("nak");
}

static void
tpacket(const char *s, struct tftphdr *tp, int n)
{
	static const char *opcodes[] =
	   { "#0", "RRQ", "WRQ", "DATA", "ACK", "ERROR", "OACK" };
	char *cp, *file;
	u_short op = ntohs((u_short)tp->th_opcode);

	if (op < RRQ || op > ERROR)
		printf("%s opcode=%x ", s, op);
	else
		printf("%s %s ", s, opcodes[op]);
	switch (op) {

	case RRQ:
	case WRQ:
		n -= 2;
		file = cp = (char *) &(tp->th_stuff);
		cp = strchr(cp, '\0');
		printf("<file=%s, mode=%s>\n", file, cp + 1);
		break;

	case DATA:
		printf("<block=%d, %d bytes>\n", ntohs(tp->th_block), n - 4);
		break;

	case ACK:
		printf("<block=%d>\n", ntohs(tp->th_block));
		break;

	case ERROR:
		printf("<code=%d, msg=%s>\n", ntohs(tp->th_code), tp->th_msg);
		break;
	}
}

struct timeval tstart;
struct timeval tstop;

static void
startclock(void)
{
	(void)gettimeofday(&tstart, NULL);
}

static void
stopclock(void)
{

	(void)gettimeofday(&tstop, NULL);
}

static void
printstats(const char *direction, unsigned long amount)
{
	double delta;

	delta = (tstop.tv_sec+(tstop.tv_usec/100000.0)) -
		(tstart.tv_sec+(tstart.tv_usec/100000.0));
	if (verbose) {
	  printf("%s %lu bytes in %.1f seconds", direction, amount, delta);
	  printf(" [%.0f bit/s]", (amount*8.)/delta);
	  putchar('\n');
	}
}

static void
timer(int sig)
{
	int save_errno = errno;

	(void)sig;		/* Shut up unused warning */

	timeout += rexmtval;
	if (timeout >= maxtimeout) {
		printf("Transfer timed out.\n");
		errno = save_errno;
		siglongjmp(toplevel, -1);
	}
	errno = save_errno;
	siglongjmp(timeoutbuf, 1);
}
